Content.Server/GameObjects/Components/Recycling/RecyclerComponent.cs

```diff
@@ -0,0 +1,196 @@
+using Robust.Shared.Containers;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components;
+using Robust.Shared.GameObjects.Components.Map;
+using Robust.Shared.GameObjects.Systems;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Maths;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.Components.Recycling
+{
+    // TODO: Add sound and safe beep
+    [RegisterComponent]
+    public class RecyclerComponent : Component, ICollideBehavior
+    {
+#pragma warning disable 649
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+#pragma warning restore 649
+
+        public override string Name => "Recycler";
+
+        /// <summary>
+        ///     Whether or not sentient beings will be recycled
+        /// </summary>
+        [ViewVariables]
+        private bool _safe;
+
+        /// <summary>
+        ///     The percentage of material that will be recovered
+        /// </summary>
+        [ViewVariables]
+        private int _efficiency; // TODO
+
+        private bool Powered =>
+            !Owner.TryGetComponent(out PowerReceiverComponent receiver) ||
+            receiver.Powered;
+
+        private void Bloodstain()
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent appearance))
+            {
+                appearance.SetData(RecyclerVisuals.Bloody, true);
+            }
+        }
+
+        private void Clean()
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent appearance))
+            {
+                appearance.SetData(RecyclerVisuals.Bloody, false);
+            }
+        }
+
+        private bool CanGib(IEntity entity)
+        {
+            return entity.HasComponent<SpeciesComponent>() &&
+                   !_safe &&
+                   Powered;
+        }
+
+        private bool CanRecycle(IEntity entity, [MaybeNullWhen(false)] out ConstructionPrototype prototype)
+        {
+            prototype = null;
+
+            var constructionSystem = EntitySystem.Get<ConstructionSystem>();
+            var entityId = entity.MetaData.EntityPrototype?.ID;
+
+            if (entityId == null ||
+                !constructionSystem.CraftRecipes.TryGetValue(entityId, out prototype))
+            {
+                return false;
+            }
+
+            return Powered;
+        }
+
+        private void Recycle(IEntity entity)
+        {
+            // TODO: Prevent collision with recycled items
+            if (CanGib(entity))
+            {
+                entity.Delete(); // TODO: Gib
+                Bloodstain();
+                return;
+            }
+
+            if (!CanRecycle(entity, out var prototype))
+            {
+                return;
+            }
+
+            var recyclerPosition = Owner.Transform.MapPosition;
+            var lastStep = prototype.Stages[^2].Forward as ConstructionStepMaterial;
```

Comment:
This seems like a really bad way to estimate a recycler output?

Reply:
Is that not how the construction system does it? https://github.com/space-wizards/space-station-14/blob/master/Content.Server/GameObjects/EntitySystems/ConstructionSystem.cs#L162-L197

Reply:
The difference is that the construction systems knows for certain "there are two stages so we know which stage is the base and only material". You don't know that for certain here.